Pazi does not build on OSX

The signal handler relies on the `__errno_location` function and `sa_restorer` field in libc, which isn't present in OSX apparently:

```
error[E0425]: cannot find function `__errno_location` in module `libc`
   --> /Users/lcm/.cargo/registry/src/github.com-1ecc6299db9ec823/pazi-0.1.0/src/main.rs:560:28
    |
560 |         let errno = *libc::__errno_location();
    |                            ^^^^^^^^^^^^^^^^ not found in `libc`

error[E0425]: cannot find function `__errno_location` in module `libc`
   --> /Users/lcm/.cargo/registry/src/github.com-1ecc6299db9ec823/pazi-0.1.0/src/main.rs:565:38
    |
565 |         let sigaction_errno = *libc::__errno_location();
    |                                      ^^^^^^^^^^^^^^^^ not found in `libc`

error[E0560]: struct `libc::sigaction` has no field named `sa_restorer`
   --> /Users/lcm/.cargo/registry/src/github.com-1ecc6299db9ec823/pazi-0.1.0/src/main.rs:544:13
    |
544 |             sa_restorer: None,
    |             ^^^^^^^^^^^ `libc::sigaction` does not have this field
    |
    = note: available fields are: `sa_sigaction`, `sa_mask`, `sa_flags`

error: aborting due to 5 previous errors

Some errors occurred: E0425, E0560.
For more information about an error, try `rustc --explain E0425`.
error: failed to compile `pazi v0.1.0`, intermediate artifacts can be found at `/var/folders/2x/y0sqffds2js22r895qsl3q9m0000gn/T/cargo-installqwCB3A`

Caused by:
  Could not compile `pazi`.
```

Happy to throw together a patch if you can point me at the correct way to do this kind of signal handling (or even just to move it into a linux specific function)

--- a/.travis.yml
+++ b/.travis.yml
@@ -6,6 +6,9 @@ branches:
 rust:
   - stable
   - nightly
+os:
+  - linux
+  - osx
 
 addons:
   apt:
@@ -13,6 +16,12 @@ addons:
     - zsh
     - bash
     - python3
+  homebrew:
+    packages:
+    - zsh
+    - bash
+    - python3
+    - gimme
 
 env:
   global:

--- a/src/main.rs
+++ b/src/main.rs
@@ -522,10 +522,17 @@ fn intercept_ctrl_c() -> Result<(), ()> {
     // That way, when ctrl-c sends a SIGINT, the only process
     // to receive it is Pazi.
     //
+    unsafe fn get_errno() -> *mut libc::c_int {
+        #[cfg(target_os = "linux")]
+        return libc::__errno_location();
+        #[cfg(target_os = "macos")]
+        return libc::__error();
+    }
+
     unsafe {
         // Create a new process group with this process in it.
         let setpgid_res = libc::setpgid(0, 0);
-        let errno = *libc::__errno_location();
+        let errno = *get_errno();
         if setpgid_res != 0 {
             debug!("Got {} from setpgid with errno {}", setpgid_res, errno);
             return Err(());
@@ -543,6 +550,7 @@ fn intercept_ctrl_c() -> Result<(), ()> {
             sa_sigaction: libc::SIG_IGN,
             sa_mask: std::mem::zeroed(),
             sa_flags: 0,
+            #[cfg(target_os = "linux")]
             sa_restorer: None,
         };
         // Place to save old SIGTTOU handler
@@ -550,7 +558,7 @@ fn intercept_ctrl_c() -> Result<(), ()> {
 
         // Ignore SIGTTOU and save previous action
         let sigaction_res = libc::sigaction(libc::SIGTTOU, &ignore_action, &mut old_action);
-        let errno = *libc::__errno_location();
+        let errno = *get_errno();
         if sigaction_res != 0 {
             debug!("Got {} from sigaction with errno {}", sigaction_res, errno);
             return Err(());
@@ -559,12 +567,12 @@ fn intercept_ctrl_c() -> Result<(), ()> {
         // Make our process group the foreground process group
         // (giving us access to stdin, etc)
         let tcsetpgrp_res = libc::tcsetpgrp(libc::STDIN_FILENO, pgrp);
-        let errno = *libc::__errno_location();
+        let errno = *get_errno();
 
         // Put the old SIGTTOU signal handler back
         // We try to do this even if tcsetpgrp failed!
         let sigaction_res = libc::sigaction(libc::SIGTTOU, &old_action, std::ptr::null_mut());
-        let sigaction_errno = *libc::__errno_location();
+        let sigaction_errno = *get_errno();
 
         // Handle tcsetpgrp and sigaction errors
         if tcsetpgrp_res != 0 || sigaction_res != 0 {

--- a/tests/src/integ.rs
+++ b/tests/src/integ.rs
@@ -19,7 +19,7 @@ fn it_jumps() {
 
 fn it_jumps_shell(shell: &Shell) {
     let tmpdir = TempDir::new("pazi_integ").unwrap();
-    let root = tmpdir.path();
+    let root = tmpdir.path().canonicalize().unwrap();
     let mut h = HarnessBuilder::new(&root, &Pazi, shell).finish();
     let slash_tmp_path = root.join("tmp");
     let slash_tmp = slash_tmp_path.to_string_lossy();
@@ -40,7 +40,7 @@ fn it_jumps_to_exact_directory() {
 
 fn it_jumps_to_exact_directory_shell(shell: &Shell) {
     let tmpdir = TempDir::new("pazi_integ").unwrap();
-    let root = tmpdir.path();
+    let root = tmpdir.path().canonicalize().unwrap();
     let mut h = HarnessBuilder::new(&root, &Pazi, shell).finish();
     let slash_tmp_path = root.join("tmp");
     let slash_tmp = slash_tmp_path.to_string_lossy();
@@ -49,6 +49,7 @@ fn it_jumps_to_exact_directory_shell(shell: &Shell) {
 
     h.create_dir(&unvisited_dir);
     h.visit_dir(&slash_tmp);
+
     assert_eq!(h.jump("asdf"), unvisited_dir);
 }
 
@@ -63,7 +64,7 @@ fn it_jumps_to_more_frecent_items() {
 
 fn it_jumps_to_more_frecent_items_shell(shell: &Shell) {
     let tmpdir = TempDir::new("pazi_integ").unwrap();
-    let root = tmpdir.path();
+    let root = tmpdir.path().canonicalize().unwrap();
     let mut h = HarnessBuilder::new(&root, &Pazi, shell).finish();
     let a_dir_path = root.join("a/tmp");
     let b_dir_path = root.join("b/tmp");
@@ -100,7 +101,7 @@ fn it_imports_from_fasd() {
 
 fn it_imports_from_fasd_shell(shell: &Shell) {
     let tmpdir = TempDir::new("pazi_integ").unwrap();
-    let root = tmpdir.path();
+    let root = tmpdir.path().canonicalize().unwrap();
 
     {
         let mut fasd = HarnessBuilder::new(&root, &Fasd, shell).finish();
@@ -132,7 +133,7 @@ fn it_ignores_dead_dirs_on_cd() {
 
 fn it_ignores_dead_dirs_on_cd_shell(shell: &Shell) {
     let tmpdir = TempDir::new("pazi_integ").unwrap();
-    let root = tmpdir.path();
+    let root = tmpdir.path().canonicalize().unwrap();
     let mut h = HarnessBuilder::new(&root, &Pazi, shell).finish();
 
     h.create_dir(&root.join("1/tmp").to_string_lossy());
@@ -159,7 +160,7 @@ fn it_prints_list_on_lonely_z() {
 
 fn it_prints_list_on_lonely_z_shell(shell: &Shell) {
     let tmpdir = TempDir::new("pazi_integ").unwrap();
-    let root = tmpdir.path();
+    let root = tmpdir.path().canonicalize().unwrap();
     let mut h = HarnessBuilder::new(&root, &Pazi, shell).finish();
 
     h.create_dir(&root.join("1/tmp").to_string_lossy());
@@ -178,7 +179,7 @@ fn it_prints_list_on_lonely_z_shell(shell: &Shell) {
 #[test]
 fn it_handles_existing_bash_prompt_command() {
     let tmpdir = TempDir::new("pazi_integ").unwrap();
-    let root = tmpdir.path();
+    let root = tmpdir.path().canonicalize().unwrap();
     let prompt_cmd = r#"
 MY_PROMPT=1
 PROMPT_COMMAND='printf -v MY_PROMPT_OUT "\033k%s\033\\" "${MY_PROMPT}"'
@@ -210,7 +211,7 @@ fn it_handles_help_output() {
 
 fn it_handles_help_output_shell(shell: &Shell) {
     let tmpdir = TempDir::new("pazi_integ").unwrap();
-    let root = tmpdir.path();
+    let root = tmpdir.path().canonicalize().unwrap();
     let mut h = HarnessBuilder::new(&root, &Pazi, shell).finish();
     let help1 = h.run_cmd("pazi --help && echo $?");
     let help2 = h.run_cmd("z -h && echo $?");
@@ -235,7 +236,7 @@ fn it_handles_things_that_look_like_subcommands() {
 
 fn it_handles_things_that_look_like_subcommands_shell(shell: &Shell) {
     let tmpdir = TempDir::new("pazi_integ").unwrap();
-    let root = tmpdir.path();
+    let root = tmpdir.path().canonicalize().unwrap();
     let mut h = HarnessBuilder::new(&root, &Pazi, shell).finish();
 
     // map of <DirectoryName, JumpTarget>
@@ -251,11 +252,11 @@ fn it_handles_things_that_look_like_subcommands_shell(shell: &Shell) {
     .collect();
 
     for (dir, jump) in map {
-        let dirname = root.join(dir).into_os_string().into_string().unwrap();
-        h.create_dir(&dirname);
-        h.visit_dir(&dirname);
+        let dir_name = root.join(dir).into_os_string().into_string().unwrap();
+        h.create_dir(&dir_name);
+        h.visit_dir(&dir_name);
         h.visit_dir(&root.to_string_lossy());
-        assert_eq!(h.jump(jump), dirname);
-        h.delete_dir(&dirname);
+        assert_eq!(h.jump(jump), dir_name);
+        h.delete_dir(&dir_name);
     }
 }

